tools/cpp/CROSSTOOL.tpl

```diff
@@ -990,12 +978,180 @@ toolchain {
+        flag: "-D__TIME__=\"redacted\""
+      }
+    }
+  }
+
+  feature {
+    name: 'treat_warnings_as_errors'
+    flag_set {
+      action: 'c-compile'
+      action: 'c++-compile'
+      flag_group {
+        flag: "/WX"
+      }
+    }
+  }
+
+  ##
+  # C++ exceptions
+  #
+  # 'default_exceptions' is in effect automatically when user does not choose
+  # between 'use_exceptions' and 'no_exceptions'. It will enable C++ exceptions,
+  # which is the default behaviour for most C++ compilers. Do not enable/disable
+  # this feature directly.
+  #
+  # If you enable 'use_exceptions' or 'no_exceptions' globally and need to switch
+  # to the opposite feature for one target, you need to first disable the
+  # original feature due to limitation of CROSSTOOL.
+  #
+  # E.g. cc_library(... features = ["-no_exceptions", "use_exceptions"])
+  #
+  # These two features are marked with "provides: 'cpp_exceptions'" to make sure
+  # that user can never enable both of them at once for any target.
+
+  feature {
+    name: 'default_exceptions'
+    enabled: true
+    flag_set {
+      action: 'c-compile'
+      action: 'c++-compile'
+      with_feature: {
+        not_feature: 'no_exceptions'
+        not_feature: 'use_exceptions'
+      }
+      flag_group {
+        flag: "/D_HAS_EXCEPTIONS=1"
+        flag: "/EHsc"
+      }
+    }
+  }
+
+  feature {
+    name: 'use_exceptions'
+    provides: 'cpp_exceptions'
+    flag_set {
+      action: 'c-compile'
+      action: 'c++-compile'
+      with_feature: {
+        not_feature: 'no_exceptions'
+      }
+      flag_group {
+        flag: "/D_HAS_EXCEPTIONS=1"
+        flag: "/EHsc"
+      }
+    }
+  }
+
+  feature {
+    name: 'no_exceptions'
+    provides: 'cpp_exceptions'
+    flag_set {
+      action: 'c-compile'
+      action: 'c++-compile'
+      with_feature: {
+        not_feature: 'use_exceptions'
+      }
+      flag_group {
+        flag: "/D_HAS_EXCEPTIONS=0"
+        flag: "/EHs-c-"
+        flag: "/wd4577" # Suppress 'noexcept used with no exception handling mode specified'
+      }
+    }
+  }
+
+  # Trade slower build time for smaller binary
+  feature {
+    name: 'smaller_binary'
+    enabled: true
+    flag_set {
+      action: 'c-compile'
+      action: 'c++-compile'
+      with_feature: {
+        feature: 'opt'
+      }
+      flag_group {
+        flag: "/Gy" # Enable function-level linking (-ffunction-sections)
+        flag: "/Gw" # Optimize global data (-fdata-sections)
+      }
+    }
+    flag_set {
+      action: 'c++-link-executable'
+      action: 'c++-link-dynamic-library',
+      action: 'c++-link-nodeps-dynamic-library'
+      with_feature: {
+        feature: 'opt'
+      }
+      flag_group {
+        flag: '/OPT:ICF' # Fold identical functions
+        flag: '/OPT:REF' # Eliminate unreferenced functions and data
+      }
+    }
+  }
+
+  # Suppress warnings that most users do not care
+  feature {
+    name: 'ignore_noisy_warnings'
+    enabled: true
```

Comment:
I remember if we set a feature to `enabled: true`, it cannot be turned off by `--feautres=-<feature name>`, @mhlopko is that still true?

Reply:
From my experimentation, `enabled: true` can still be disabled with `--features=-*` while `implies:` will prevent disabling completely (otherwise error thrown).

Reply:
Thanks! Then LGTM!